posts/2021-07-21-building-on-atomic-reference.adoc

```diff
@@ -0,0 +1,208 @@
+---
+layout: post
+title: "Building on Atomic Reference"
+# Do NOT change the categories section
+categories: blog
+author_picture: https://avatars3.githubusercontent.com/joe-chacko
+author_github: https://github.com/joe-chacko
+seo-title: Building on AtomicReference - OpenLiberty.io
+seo-description: Java's AtomicReference does not guarantee that an update function gets called exactly once. We describe a helpful, thread-safe mechanism that builds on AtomicReference to provide such a guarantee. This can be used for lazy initialization, clean-up, or any state change processing that is to happen exactly once.
+blog_description: "Java's AtomicReference does not guarantee that an update function gets called exactly once. We describe a helpful, thread-safe mechanism that builds on AtomicReference to provide such a guarantee. This can be used for lazy initialization, clean-up, or any state change processing that is to happen exactly once."
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+additional_authors:
+- name: Neil Richards
+  github: https://github.com/ngmr
+  image: https://avatars0.githubusercontent.com/ngmr
+---
+= Building on AtomicReference
+Joe Chacko <https://github.com/joe-chacko>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+The https://devdocs.io/openjdk/java.base/java/util/concurrent/atomic/atomicreference[`AtomicReference`] and other `Atomic...` classes in the `java.util.concurrent` package allow multiple threads to read and update values safely.
+Yet `AtomicReference` alone is of limited use in writing thread-safe code.
+This article describes a helpful mechanism to provide a more useful thread-safe reference.
+
+== Something is rotten in the state transition
+Consider the https://devdocs.io/openjdk~15/java.base/java/util/concurrent/atomic/atomicreference#updateAndGet(java.util.function.UnaryOperator)[`updateAndGet()`] method.
+This allows many threads to contend to update the value.
+For example, one might write something like this:
+[source, java]
+----
+    private AtomicReference<Widget> ref;
+
+    public Widget getWidget() {
+        return ref.updateAndGet(value -> value == null ? new Widget() : value);
+    }
+----
+When `getWidget()` is first invoked, it will create a new `Widget`, but later calls will find the existing one.
+
+But there is a trap here for the unwary developer.
+The contract of the `updateAndGet()` method describes some restrictions on the update function: --
```

Comment:
would it be helpful to embed a link to the javadoc here?
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html
not sure if that's the best source, but whatever is most relevant for your reference

Reply:
Already linked in the previous para. I don't see how to add links inside a source block if that's what you mean.